core/modules/parsers/videoparser.js

```diff
@@ -6,29 +6,161 @@ module-type: parser
 The video parser parses a video tiddler into an embeddable HTML element
 
 \*/
-(function(){
+(function () {
 
 /*jslint node: true, browser: true */
 /*global $tw: false */
 "use strict";
 
-var VideoParser = function(type,text,options) {
+var VideoParser = function (type, text, options) {
 	var element = {
+		type: "element",
+		tag: "video",
+		attributes: {
+			controls: { type: "string", value: "controls" },
+			style: { type: "string", value: "width: 100%; object-fit: contain" },
+			preload: { type: "string", value: "auto" },
+			class: { type: "string", value: "tw-video-element" }
+		}
+	};
+
+	if (options._canonical_uri) {
+		element.children = [{
 			type: "element",
-			tag: "video",
+			tag: "source",
 			attributes: {
-				controls: {type: "string", value: "controls"},
-				style: {type: "string", value: "width: 100%; object-fit: contain"}
+				src: { type: "string", value: options._canonical_uri },
+				type: { type: "string", value: type }
 			}
-		},
-		src;
-	if(options._canonical_uri) {
-		element.attributes.src = {type: "string", value: options._canonical_uri};
-	} else if(text) {
-		element.attributes.src = {type: "string", value: "data:" + type + ";base64," + text};
+		}];
+	} else if (text) {
+		element.attributes.src = { type: "string", value: "data:" + type + ";base64," + text };
+	}
+
+	if ($tw.browser) {
+		const processedVideos = new WeakMap();
+		const bufferThreshold = 0.1; // 10% buffered before play
+		
+		$tw.hooks.addHook("th-page-refreshed", function() {
+			Array.from(document.getElementsByClassName("tw-video-element")).forEach(function(video) {
+				if (processedVideos.has(video)) return;
+				
+				// Create loading overlay
+				const overlay = document.createElement('div');
+				overlay.style.cssText = 'position:absolute;top:0;left:0;width:100%;height:100%;background:rgba(0,0,0,0.5);display:flex;align-items:center;justify-content:center;color:white;';
+				overlay.innerHTML = 'Loading 0%';
```

Comment:
hmmm, `innerHTML` -- I am not sure if a "parser" should modify the DOM structure. Ususally widgets need to be used to create the DOM